### ISSUE TITLE
Add th:uerrors for unformatted error messages

### ISSUE THREAD

**Issue by jmitash** [thymeleaf/thymeleaf-spring#153]
I am working with a framework where non-ASCII characters, like á, é, í, ó, ú, are represented by their escaped HTML versions, `&aacute;`, `&eacute;`, `&iacute;`, `&oacute;` , `&uacute;`. I am trying to get error messages to display appropriately with these escaped strings, but `th:errors` escapes them a second time anyway. As a result, I get error messages like this:

> Por favor ingrese el c&amp;oacute;digo postal.

My current solution is to replace the `th:errors` with a `th:each` followed by a `th:utext` in order to get the raw error messages. While this works, it's verbose and a lot less convenient than a `th:uerrors` would be. 

I'm proposing `th:uerrors` which would behave exactly the same as `th:errors` but leave the text unformatted as `th:utext` does.

**Comment by danielfernandez** [comment]
Having an entire new attribute processor for this seems a bit overkill to me, at least at first thought, given the existence of the `#fields` expression object, with methods `allErrors()`, `errors()` or `globalErrors()` which will allow you to obtain the list of the specific type of errors you need and easily iterate it and output each of the individual errors in the way you need (for instance, using `th:utext`).

Given yours is a quite specific need, I think it can be handled this way (in the end, all `th:errors` does is add some line breaks). Is there any reason why this wouldn't work for you?

**Comment by jmitash** [comment]
> Given yours is a quite specific need, I think it can be handled this way (in the end, all `th:errors` does is add some line breaks). Is there any reason why this wouldn't work for you?

While it would work, it is less than ideal for my situation. I'll try to avoid the specifics of what I'm working with, but everything revolves around an e-commerce framework. The framework is built to be highly extensible and modifiable. The site we are building on top of the framework is designed to serve as a demonstration of the framework's capabilities (and therefore must also be easily extensible). But at the same time, the site is meant to be a starting point for developers who want to get an online store up and running.

Since we expect other people to use the site to build their own, it's very important that the site has a clean source and is easy to work with. One of the features of the framework is it's internationalization. Since the site is a reflection of the framework, it's only natural that site be internationalized as well. The end goal is that the site has absolutely no static messages, and that every text on the site can be replaced with any other language. That's the reason that unformatted text is important to us.

When the site was started, we were working with only one language (English). We used `th:errors` thinking it would serve the right purpose. Later on we tested other languages, at which point we realized that `th:errors` didn't leave the raw message that we needed. It was clear that `th:utext` would give unformatted text because it has the accompanying tag `th:text`. With `th:errors`, It didn't cross my mind that it would escape the text. I feel like a developer building off the site could easily make the same mistake and not realize `th:errors` is unformatted. Even then, I would argue that in that case, it's not obvious that you would need to use a `th:each`, `#fields`, and `th:utext` to get the desired output. A `th:uerrors` would be much more intuitive.

I started replacing fields such as
```HTML
<span class="text-danger" th:errors="*{customer.emailAddress}"></span>
```
with
```HTML
<span class="text-danger" th:each="err : ${#fields.errors('customer.emailAddress')}"
      th:utext="${err}"></span>
```

as you suggested. I vaguely mentioned it in the original comment as well. While this would have worked, I didn't feel like it was a clean solution (keep in mind a major goal of the site is to keep the source clean). There are about 450 references to `th:errors` on the site (and we have more in other repositories that also require i18n). Writing out

```HTML
th:each="err : ${#fields.errors('customer.emailAddress')}" th:utext="${err}"
```
instead of
```HTML
th:uerrors="*{customer.emailAddress}"
```

is more verbose, less readable, and more error prone.

You mentioned that this is a specific need for me, so I'd like to take a step back and mention how others can benefit as well. It might seem unique that we need unformatted text throughout the whole project, but this could just as easily apply to anyone else trying to internationalize their site. Even if they are only supporting languages that do not require escaped HTML characters, it is still beneficial to be able to serve unformatted text. For example, a certain word of the error message might benefit from emphasis:

> You must enter a shipping address **and** a billing address.

Again, they could replace that `th:errors` with the `th:each`, `#field.errors()`, and `th:utext`, but that leads to inconsistencies. Some fields with use `th:errors`, some will use the above, and in both cases the developer will have to be aware of whether the corresponding message has characters that need to be left unescaped. They could replace all `th:errors`, but then they would suffer because that solution is

>more verbose, less readable, and more error prone.

You also mentioned you think another processor is overkill. I already did my best to explain the benefits, so I'll look at the downsides. As far as implementation goes, these are the changes I would think are necessary

- `SpringUErrorClassTagProcessor` to process the new tag
- `SpringErrorClassTagProcessor` to remove the the logic from and put it into the...
- `AbstractSpringErrorClassTagProcessor` to hold shared logic of the error processors
- `SpringStandardDialect` to add the new processor

Hardly any new code would be introduced, public interfaces would be unaffected, and no complex logic (or any notable logic) would be added. Implementation-wise, I think this would be extremely simple.

And besides just implementing it and documenting it, I really don't see any other downsides. Sorry for giving you an essay to read, but I am pretty passionate that this would be beneficial and wanted you to see my point of view. If you want me to provide the implementation I'd be happy to (on my own time). Appreciate your time.

**Comment by danielfernandez** [comment]
Yes, I know Broadleaf of course. You guys do an awesome job.

I understand what you say, and I've investigated a bit and seen that Spring's own JSP "errors" tag does actually have an `htmlEscape` attribute which allows the error messages to be output unescaped. So it's true that `th:uerrors` would fit as a match for that functionality.

I'm still a bit worried about a thing, and it is the possibility of code injection. If error messages are defined to contain the input value, and this contains malicious code, outputting it directly without escaping would result in trouble. Other attribute processors such as `th:utext` forbid expression preprocessing and the use of request parameters precisely to protect against this, but in this case Thymeleaf cannot really know what was in the error message pattern originally (only the result of its rendering), so identifying such cases would be probably impossible.

This said, none of the default messages defined for the Hibernate Validator (`ValidatorMessages.properties`) apparently include the input value, and Spring itself does not perform any other message substitutions other than the name (not the *value*) of the field (with `{0}`). There are cases in which the value would be output, like with data type conversions, but that is because in those cases there are no default validation messages and instead what is output is the message of a conversion exception. So we can assume this type of error messages would be customised by application developers when possible.

So under the Spring umbrella (and unless I'm missing something) it seems that creating validation messages including the input value would not be possible unless people customise the JSR349/380 message interpolation mechanism — and thus this use case probably could be deemed *safe by default* (though it should be warned against at the docs anyway, of course).

Is this aspect something you've considered? Do you use the Spring/JSR349/JSR380 standard infrastructure for these validation/error messages or have you got a custom mechanism in place?

**Comment by jmitash** [comment]
> You guys do an awesome job.

Thanks, it goes without saying we greatly appreciate your team's work too. I tried to avoid mentioning the company here because I wasn't sure the Contribution guide applied here on GitHub as well.

I hadn't considered code injection a problem to be honest. Especially because as you discussed, we don't use the value in any of the messages, so it didn't really raise any concerns for me specifically. We are using the standard Spring provided validators, however they can be overridden, grouping us into the "safe by default" category you mentioned.

I am not very concerned that someone would override the validator using our starter site (and expose themselves to code injection). I don't see much to benefit from doing so. I imagine if they required their own validator they would be building the site on their own, or integrating Broadleaf into their existing site. Even then, they would have to be aware of the dangers of overriding a validator with their own in the first place.

**Comment by danielfernandez** [comment]
This has been implemented and is already in `3.0.8-SNAPSHOT`

**Comment by jmitash** [comment]
Much appreciated, thanks for the help Daniel.
